llvm/tools/llvm-cas-fuzzer/llvm-cas-fuzzer.cpp

```diff
@@ -0,0 +1,247 @@
+      DataLength = DataLength % 1024;
+
+    if (ForceKill) {
+      Settings |= FORK;
+      Settings |= CHECK_TERMINATION;
+    }
+  }
+
+  void init() {
+    NumShards = OptNumShards ? OptNumShards : 8;
+    NumChildren = OptNumChildren ? OptNumChildren : 4;
+    TreeDepth = OptTreeDepth ? OptTreeDepth : 12;
+    DataLength = OptDataLength ? OptDataLength : 128;
+  }
+
+  void appendCommandLineOpts(std::vector<std::string> &Cmd) {
+    Cmd.push_back("--num-shards=" + utostr(NumShards));
+    Cmd.push_back("--num-children=" + utostr(NumChildren));
+    Cmd.push_back("--tree-depth=" + utostr(TreeDepth));
+    Cmd.push_back("--data-length=" + utostr(DataLength));
+  }
+
+  void dump() {
+    llvm::errs() << "## Configuration:"
+                 << " Fork: " << (bool)(Settings & FORK)
+                 << " Kill: " << (bool)(Settings & CHECK_TERMINATION)
+                 << " Prune: " << (bool)(Settings & PRUNE_CAS)
+                 << " NumShards: " << (unsigned)NumShards
+                 << " TreeDepth: " << (unsigned)TreeDepth
+                 << " NumChildren: " << (unsigned)NumChildren
+                 << " DataLength: " << (unsigned)DataLength << "\n";
+  }
+};
+
+static void parseOptions(int Argc, char **Argv) {
+  Argv0 = Argv[0];
+
+  for (int I = 0; I < Argc; ++I) {
+    StringRef Arg = Argv[I];
+    // option must start with `--`.
+    if (!Arg.consume_front("--"))
+      continue;
+
+    // flags.
+    if (Arg == "gen-data") {
+      GenData = true;
+      continue;
+    }
+    if (Arg == "print-config") {
+      PrintConfig = true;
+      continue;
+    }
+    if (Arg == "force-kill") {
+      ForceKill = true;
+      continue;
+    }
+
+    // options that take an argument.
+    auto Opt = Arg.split('=');
+    if (Opt.first == "cas-path")
+      CASPath = Opt.second.str();
+    else if (Opt.first == "num-shards")
+      to_integer(Opt.second, OptNumShards);
+    else if (Opt.first == "num-children")
+      to_integer(Opt.second, OptNumChildren);
+    else if (Opt.first == "tree-depth")
+      to_integer(Opt.second, OptTreeDepth);
+    else if (Opt.first == "data-length")
+      to_integer(Opt.second, OptDataLength);
+  }
+}
+
+// fill the CAS with random data of specified tree depth and children numbers.
+static void fillData(ObjectStore &CAS, const Config &Conf) {
+  ExitOnError ExitOnErr("llvm-cas-fuzzer fill data: ");
+  DefaultThreadPool ThreadPool(hardware_concurrency());
+  for (size_t I = 0; I != Conf.NumShards; ++I) {
+    ThreadPool.async([&] {
+      std::vector<ObjectRef> Refs;
+      for (unsigned Depth = 1; Depth <= Conf.TreeDepth; ++Depth) {
+        unsigned NumNodes = (Conf.TreeDepth - Depth) * Conf.NumChildren + 1;
+        std::vector<ObjectRef> Created;
+        Created.reserve(NumNodes);
+        ArrayRef<ObjectRef> PreviouslyCreated(Refs);
+        for (unsigned I = 0; I < NumNodes; ++I) {
+          std::vector<char> Data(Conf.DataLength);
+          getRandomBytes(Data.data(), Data.size());
```

Comment:
We should probably have a seed and print it on failure to have a chance of reproducing.

Reply:
In my opinion, the actual data that get inserted is absolutely not important and you cannot reproduce even with a stable seed anyway. We are fuzzing to find concurrency issues, and the random data is just to make sure an actual insert is happen, rather than just a duped entry and CAS never grows.

The way to debug is the looking at the resulting CAS directory and looking at CAS log to figure out how we get here. TSAN build is also recommended.

Reply:
I agree that even reproducing with stable seed is not 100% guaranteed, but I don't think the exact data is unimportant. Admittedly, I don't know the implementation details here, but I assume the code paths taken when inserting a new blob vs previously-seen blob must be different. Why not at least take that part out of the reproducibility equation?

Reply:
The idea here is always inserting new data as it creates writes that can potentially break CAS. maybe the fuzzer should sometime insert same data as well, but I need to think if that is interesting at all.